ql/src/java/org/apache/hadoop/hive/ql/txn/compactor/metrics/DeltaFilesMetricReporter.java

```diff
@@ -0,0 +1,296 @@
+import org.apache.hadoop.hive.common.metrics.common.MetricsVariable;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.ql.io.AcidDirectory;
+import org.apache.hadoop.hive.ql.io.AcidUtils;
+
+import org.apache.hadoop.hive.shims.HadoopShims;
+import org.apache.hadoop.mapred.JobConf;
+import org.apache.hive.common.util.Ref;
+import org.apache.tez.common.counters.TezCounters;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.util.Comparator;
+
+import java.util.Map;
+import java.util.Queue;
+import java.util.concurrent.Executors;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.PriorityBlockingQueue;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Stream;
+
+import static org.apache.hadoop.hive.metastore.metrics.MetricsConstants.COMPACTION_NUM_DELTAS;
+import static org.apache.hadoop.hive.metastore.metrics.MetricsConstants.COMPACTION_NUM_OBSOLETE_DELTAS;
+import static org.apache.hadoop.hive.metastore.metrics.MetricsConstants.COMPACTION_NUM_SMALL_DELTAS;
+
+/**
+ * Collects and publishes ACID compaction related metrics.
+ */
+public class DeltaFilesMetricReporter {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AcidUtils.class);
+
+  public static final String NUM_OBSOLETE_DELTAS = "HIVE_ACID_NUM_OBSOLETE_DELTAS";
+  public static final String NUM_DELTAS = "HIVE_ACID_NUM_DELTAS";
+  public static final String NUM_SMALL_DELTAS = "HIVE_ACID_NUM_SMALL_DELTAS";
+
+  private int deltasThreshold;
+  private int obsoleteDeltasThreshold;
+
+  private int maxCacheSize;
+
+  private Cache<String, Integer> deltaCache, smallDeltaCache;
+  private Cache<String, Integer> obsoleteDeltaCache;
+
+  private BlockingQueue<Pair<String, Integer>> deltaTopN, smallDeltaTopN;
+  private BlockingQueue<Pair<String, Integer>> obsoleteDeltaTopN;
+
+  private ScheduledExecutorService executorService;
+
+  private static class InstanceHolder {
+    public static DeltaFilesMetricReporter instance = new DeltaFilesMetricReporter();
+  }
+
+  private DeltaFilesMetricReporter() {
+  }
+
+  public static DeltaFilesMetricReporter getInstance() {
+    return InstanceHolder.instance;
+  }
+
+  public static synchronized void init(HiveConf conf){
+    getInstance().configure(conf);
+  }
+
+  public void submit(TezCounters counters) {
+    updateMetrics(NUM_OBSOLETE_DELTAS,
+        obsoleteDeltaCache, obsoleteDeltaTopN, obsoleteDeltasThreshold, counters);
+    updateMetrics(NUM_DELTAS,
+        deltaCache, deltaTopN, deltasThreshold, counters);
+    updateMetrics(NUM_SMALL_DELTAS,
+        smallDeltaCache, smallDeltaTopN, deltasThreshold, counters);
+  }
+
+  public static Pair<Integer, Integer> getNumDeltas(AcidDirectory dir, long checkThresholdInSec,
```

Comment:
style-related: it's a little weird that getNumDeltas returns Pair<Integer, Integer> and getNumObsoleteDeltas returns int, and they're only ever called together. What do you think about merging these 2 functions into 1?

If you don't like that then at least, getNumDeltas needs a javadoc comment describing what it returns.

Reply:
fixed